feat: add wsproxy implementation for key fetching

sreya opened the pull request:
This PR implements the `wsproxy` variant for a key cache. Along with `jwt` PR these are the final pieces before we can start wiring everything up.

sreya committed:
feat: add wsproxy implementation for key fetching

sreya committed:
Refactor CryptoKey usage to use codersdk package

This change streamlines the CryptoKey handling by utilizing the
codersdk package, thereby reducing code duplication and potentially
simplifying maintenance efforts in the future.

sreya committed:
Refactor wsproxy to use cryptokeys interface

- Implement `cryptokeys.Keycache` interface in `CryptoKeyCache`.
- Introduce context management for graceful shutdowns.
- Simplify function signatures and improve concurrency handling.
- Ensure functions return errors when cache is closed.

sreya committed:
Refactor keycache to improve error handling

- Use `cryptokeys.ErrKeyNotFound` for missing keys
- Use `cryptokeys.ErrKeyInvalid` for invalid keys
- Replace `xerrors` for streamlined error codes

sreya committed:
prevent cache slowdown during fetches

sreya committed:
fix tests

Emyrk commented on enterprise/wsproxy/keycache.go:
```diff
@@ -0,0 +1,192 @@
+
+func (k *CryptoKeyCache) Signing(ctx context.Context) (codersdk.CryptoKey, error) {
+	k.keysMu.RLock()
+
+	if k.closed {
+		k.keysMu.RUnlock()
+		return codersdk.CryptoKey{}, cryptokeys.ErrClosed
+	}
+
+	latest := k.latest
+	k.keysMu.RUnlock()
+
+	now := k.Clock.Now()
+	if latest.CanSign(now) {
+		return latest, nil
+	}
+
+	k.keysMu.Lock()
+	defer k.keysMu.Unlock()
+
+	if k.closed {
+		return codersdk.CryptoKey{}, cryptokeys.ErrClosed
+	}
+
+	if k.latest.CanSign(now) {
+		return k.latest, nil
+	}
+
+	var err error
+	k.keys, k.latest, err = k.fetch(ctx)
+	if err != nil {
+		return codersdk.CryptoKey{}, xerrors.Errorf("fetch: %w", err)
+	}
+
+	if !k.latest.CanSign(now) {
+		return codersdk.CryptoKey{}, cryptokeys.ErrKeyNotFound
+	}
+
+	return k.latest, nil
+}
```
This looks very similar to the `dbcache` [`dbkeycache.go`](https://github.com/coder/coder/blob/0aa84b18a10382f70bda7ee4b1c3f4b6bdcd22f9/coderd/cryptokeys/dbkeycache.go#L108-L142).

Can we DRY up these keycache implementations?

sreya committed:
craft test for race condition

sreya committed:
remove gotestleak...too many unrelated errors

sreya committed:
optimize Close

spikecurtis commented on enterprise/wsproxy/keycache.go:
```diff
@@ -0,0 +1,252 @@
+	}
+
+	k.refreshCancel()
+	k.closed.Store(true)
```
Multiple calls to `Close()` could race such that they both execute.  Right now `Close()` is idempotent, but I think it's better to just make `closed` a regular boolean protected by the mutex, so we don't have to worry about it getting executed more than once.

spikecurtis commented on enterprise/wsproxy/keycache_test.go:
```diff
@@ -0,0 +1,504 @@
+}
+
+type fakeCoderd struct {
+	server *httptest.Server
```
If you made `CryptoKeyCache.client` an interface, you wouldn't need to mess around with HTTP, URLs, or JSON.  You could just directly return the keys you want to test with.

spikecurtis commented on enterprise/wsproxy/keycache.go:
```diff
@@ -0,0 +1,252 @@
+
+	now := k.Clock.Now()
+	if !latest.CanSign(now) {
+		return nil, codersdk.CryptoKey{}, cryptokeys.ErrKeyInvalid
```
Rare, but if you're fetching to get a verification key, it seems wrong to throw an error based on the signing key.

spikecurtis commented on enterprise/wsproxy/keycache.go:
```diff
@@ -0,0 +1,252 @@
+	// There's a window we must account for where the timer fires while a fetch
+	// is ongoing but prior to the timer getting reset. In this case we want to
+	// avoid double fetching.
+	if now.Sub(lastFetch) < time.Minute*10 {
```
pull 10m into a constant

Emyrk commented:
Spike's comments I think highlight the need to implement this once. With the changes https://github.com/coder/coder/pull/14928, can we reuse 1 implementation?

sreya commented:
@Emyrk I'm game to refactor this to use one implementation but honestly I'd like to get this initiative over the finish line. I'd consider the refactor tech debt which I can tackle concurrently to ensuring there are no larger issues with the overall feature while we dogfood it in dev. You chill with that?

Emyrk commented:
> @Emyrk I'm game to refactor this to use one implementation but honestly I'd like to get this initiative over the finish line. I'd consider the refactor tech debt which I can tackle concurrently to ensuring there are no larger issues with the overall feature while we dogfood it in dev. You chill with that?

I can live with it :+1:

sreya commented on enterprise/wsproxy/keycache.go:
```diff
@@ -0,0 +1,252 @@
+	keysMu    sync.RWMutex
+	keys      map[int32]codersdk.CryptoKey
+	latest    codersdk.CryptoKey
+	fetchLock sync.RWMutex
```
Honestly I haven't messed with sync.Cond that much this is super clean

sreya committed:
Refactor CryptoKeyCache to improve concurrency

- Replaced sync/atomic with sync.Mutex and sync.Cond for better
  concurrency control, avoiding frequent lock contention and reducing
  potential for data races.
- Removed separate fetch logic and integrated it into the main crypto
  key retrieval flow to handle concurrent fetches properly.
- Optimized the Close method to minimize wait time by broadcasting on
  cond to notify waiting fetches when closed.

sreya committed:
Refactor key refresh to minimize lock duration

sreya committed:
Merge branch 'main' into jon/wspkeys

sreya committed:
Remove unused interface implementation check

The unused `cryptokeys.Keycache` interface implementation check was removed from the `CryptoKeyCache` struct, as it was unnecessary and did not contribute to functionality. This streamlines the code and adheres to best practices for interface checks.

spikecurtis commented on enterprise/wsproxy/keycache.go:
```diff
@@ -0,0 +1,225 @@
+}
+
+func (k *CryptoKeyCache) key(sequence int32) (codersdk.CryptoKey, bool) {
+	if sequence == latestSequence {
```
if you're going to use a special int32 to represent latest, you might as well go whole hog and just insert the crypto key into the map under that special value, and drop `k.latest`

sreya committed:
Remove redundant CryptoKey cache field

The `latest` field in `CryptoKeyCache` was redundant and has been removed to simplify the code. Instead, the latest crypto key is now directly accessed from the `keys` map using `latestSequence`. This change reduces unnecessary state management and potential for data inconsistency.

sreya committed:
Merge branch 'main' into jon/wspkeys
